doc/Gemm.md

```diff
@@ -0,0 +1,405 @@
+
+  - [GEMM Backends](#gemm-backends)
+  - [GEMM Launcher](#gemm-launcher)
+  - [Source Code Generation](#source-code-generation)
+
+- [**GEMM Configurations**](#gemm-configurations)
+
+  - [Backend Configurations](#backend-configurations)
+  - [CMake Configurations](#cmake-configurations)
+
+- [**Common Tasks**](#common-tasks)
+
+  - [Adding a new configuration](#adding-a-new-configuration)
+  - [Adding a new kernel](#adding-a-new-kernel)
+
+# GEMM
+
+## What is GEMM?
+
+`GEMM` stands for General Matrix Multiplication and solves the equation of the form:
+
+```
+C = alpha * A * B + beta * C
+where A, B and C are matrices and alpha and beta are scalars.
+```
+
+## SYCL-BLAS GEMM Kernels
+
+`SYCL-BLAS` currently contains the following `GEMM` kernels in <src/operations/blas3/>:
+
+- `gemm_ref.hpp` - A naive, reference implementation of `GEMM` with no optimizations.
+
+- `gemm_partial_local.hpp` - Used for tall, skinny `GEMM` optimizations.
+
+- `gemm_local.hpp` - Uses local memory for increased performance on platforms that have it. Supports full vectorization.
+
+- `gemm_no_local_partial_vec.hpp` - Doesn't use local memory. Only supports partial vectorization (can only vectorize in some specific cases).
+
+- `gemm_no_local_full_vec.hpp` - Doesn't use local memory. Supports full vectorization.
+
+- `gemm_interleaved.hpp` - An alternative approach to batched `GEMM` calculations where the inputs are interleaved in contiguous memory. Uses no local memory and corresponds to HWN data layout (NWH in column major, which is what `SYCL-BLAS` uses).
+
+## Relevant CMake Variables
+
+There are several CMake variables which are specific to `GEMM` :
+
+- `NAIVE_GEMM` (Default: `OFF`) - Forces the use of the naive, reference `GEMM` implementation.
+- `GEMM_VECTORIZATION_SUPPORT` (Default: `OFF`) - Enables vectorization within the `GEMM` kernels. If `OFF` it is equivalent to passing `1` for the vector size to the `Gemm` launcher.
+- `GEMM_TALL_SKINNY_SUPPORT` (Default: `ON`) - Enables optimizations for tall, skinny matrices. Not used on all targets.
+- `BLAS_MODEL_OPTIMIZATION` - Passing a machine learning model name here (`VGG_16` or `RESNET_50`) enables optimizations for the `Gemm` sizes used in these models. Only applies to the `ARM_GPU` target.
+
+## Kernel Structure
+
+Kernels are created as partial template specializations of the `Gemm` class to optimize for more specific cases or use different features (such as using local memory). The `Gemm` class has a number of template parameters, some of which are typically used for partial specialization. The class definition (along with the naive, reference implementation) is located in `gemm_ref.hpp`.
+
+There are some small member functions which do things such as determine the number of work groups required to execute each `Gemm`. However, the actual work of each kernel is done in `Gemm::eval()`, though the approach each kernel takes within this function is slightly different. 
```

Comment:
```suggestion
There are some small member functions which do things such as determine the number of work groups required to execute each `Gemm`. However, the actual work of each kernel is done in `Gemm::eval()`, though this approach each kernel takes within this function is slightly different. 
```

I could be wrong here, but I think that's what you meant

Reply:
I don't quite understand the suggested change, but I think the second half of that sentence is probably unclear and a little pointless so I'll change it.

Reply:
I've removed this part of the sentence!